aws/src/main/python/beam_lambda/lambda_function.py

```diff
@@ -232,44 +336,90 @@ def deploy_spot_instance(script, instance_type, region_prefix, shutdown_behaviou
+            'IamInstanceProfile': {'Name': os.environ['IAM_ROLE'] },
+            'TagSpecifications': [
+                {
+                    'ResourceType': 'instance',
+                    'Tags': [
+                        {
+                            'Key': 'Name',
+                            'Value': instance_name
+                        }, {
+                            'Key': 'GitUserEmail',
+                            'Value': git_user_email
+                        }, {
+                            'Key': 'DeployType',
+                            'Value': deploy_type_tag
+                        }
+                    ]
+                }
+            ]
+        }
+        launch_specifications.append(specification)
+    spot_fleet_req = ec2.request_spot_fleet(
+        SpotFleetRequestConfig={
+            'AllocationStrategy': 'lowestPrice',
+            'TargetCapacity': 1,
+            'IamFleetRole': 'arn:aws:iam::340032650202:role/aws-ec2-spot-fleet-tagging-role',
+            'Type': 'request',
+            'InstanceInterruptionBehavior': shutdown_behaviour,
+            'LaunchSpecifications': launch_specifications
         }
     )
-    state = 'open'
-    spot_req_id = spot_req.get('SpotInstanceRequests')[0].get('SpotInstanceRequestId')
-    while state == 'open':
-        print 'Waiting for spot request id to move from open'
-        time.sleep(30)
-        spot = ec2.describe_spot_instance_requests(SpotInstanceRequestIds = [spot_req_id]).get('SpotInstanceRequests')[0]
-        state = spot.get('State')
-    if (state != 'active'):
+    status = 'pending_fulfillment'
+    state = 'submitted'
+    spot_fleet_req_id = spot_fleet_req.get('SpotFleetRequestId')
+    print 'SpotFleetRequestId is ' + spot_fleet_req_id
+    #Flow as far as I know is that state goes to submitted, then active, but isn't done until status is out of pending_fulfillment
+    while status == 'pending_fulfillment' or state == 'submitted':
+        remaining_time = context.get_remaining_time_in_millis()
+        print 'Waiting for spot fleet request id to finish pending_fulfillment - Status: ' + status + ' and State: ' + state + ' and Remaining Time (ms): ' + str(remaining_time)
+        if remaining_time <= 60000:
+            ec2.cancel_spot_fleet_requests(
+                DryRun=False,
+                SpotFleetRequestIds=[spot_fleet_req_id],
+                TerminateInstances=True
+            )
+            print 'Waiting 30 seconds to let spot fleet cancel and then shutting down due to getting too close to lambda timeout'
+            time.sleep(30)
+            exit(123)
+        else:
+            time.sleep(30)
+            spot = ec2.describe_spot_fleet_requests(SpotFleetRequestIds = [spot_fleet_req_id]).get('SpotFleetRequestConfigs')[0]
+            status = spot.get('ActivityStatus')
+            state = spot.get('SpotFleetRequestState')
+    if state != 'active' or status != "fulfilled":
         exit(1)
```

Comment:
Can we be sure at that point we do not have any dangling instances?
Like request was fulfilled, but the instance is still initializing or something else.
If something like that happens then will our AWS-idle-notification catch that?

Reply:
This is more of a failsafe - but I can add the cancel call to be safe - I think if it gets here then it failed altogether, though. But the idle check would indeed be the ultimate failsafe

Reply:
Fixed via https://github.com/LBNL-UCB-STI/beam/pull/2511/commits/749e812dead43ef7c267bec0a2c78491e3439942